pkg/volume/csi/csi_attacher.go

```diff
@@ -269,24 +269,34 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		return err
 	}
 
-	dataDir := filepath.Dir(deviceMountPath)
-	if err := os.MkdirAll(dataDir, 0750); err != nil {
-		glog.Error(log("attacher.MountDevice failed to create dir %#v:  %v", dataDir, err))
+	// Store volume metadata for UnmountDevice. Keep it around even if the
+	// driver does not support NodeStage, UnmountDevice still needs it.
+	if err = os.MkdirAll(deviceMountPath, 0750); err != nil {
+		glog.Error(log("attacher.MountDevice failed to create dir %#v:  %v", deviceMountPath, err))
 		return err
 	}
-
+	glog.V(4).Info(log("created target path successfully [%s]", deviceMountPath))
+	dataDir := filepath.Dir(deviceMountPath)
 	data := map[string]string{
 		volDataKey.volHandle:  csiSource.VolumeHandle,
 		volDataKey.driverName: csiSource.Driver,
 	}
-	if err := saveVolumeData(dataDir, volDataFileName, data); err != nil {
+	if err = saveVolumeData(dataDir, volDataFileName, data); err != nil {
 		glog.Error(log("failed to save volume info data: %v", err))
-		if err := os.RemoveAll(dataDir); err != nil {
-			glog.Error(log("failed to remove dir after error [%s]: %v", dataDir, err))
-			return err
+		if cleanerr := os.RemoveAll(dataDir); err != nil {
+			glog.Error(log("failed to remove dir after error [%s]: %v", dataDir, cleanerr))
 		}
 		return err
 	}
+	defer func() {
+		if err != nil {
+			// clean up metadata
+			glog.Errorf(log("attacher.MountDevice failed: %v", err))
```

Comment:
Also should all the early error returns set err first so that the correct error will be logged here?  For example L311, 322, 331, 336, 345

Reply:
Fixed 322, 331 336 and 345. I can't see anything wrong in 311.

Reply:
My question regarding 311, 332, 337: is there a reason to log almost the same message first before returning the error, which will then be logged again in the defer?

Reply:
I see, I removed these error logs.